### PR TITLE
KAFKA-16483: migrate DeleteOffsetsConsumerGroupCommandIntegrationTest to use ClusterTestExtensions

### PR DESCRIPTION
**Pull request by FrankYang0529**
By using ClusterTestExtensions, `DeleteOffsetsConsumerGroupCommandIntegrationTest` get away from `KafkaServerTestHarness` dependency.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Could we define `ClusterTestDefaults` at class-level?

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for adopting new infra. however it seems the test get unstable. please help us dig in it.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Yes, updated it. Thank you.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: could you make sure all services get closed?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Can we avoid using scala code here?

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updated PR. two small comments left.

**Comment by chia7712** [comment]
@FrankYang0529 Could you reduce the partition number of offsets topic? It seems the timeout is caused by that coordinator is waiting for the offset partition, and our CI could be too busy to complete the assignments.

**Comment by FrankYang0529** [comment]
> @FrankYang0529 Could you reduce the partition number of offsets topic? It seems the timeout is caused by that coordinator is waiting for the offset partition, and our CI could be too busy to complete the assignments.

Hi @chia7712, thanks for the suggestion. I have set `offsets.topic.num.partitions` as `1` on `ClusterTestDefaults`. Hope it works fine.

**Comment by FrankYang0529** [comment]
> > @FrankYang0529 Could you reduce the partition number of offsets topic? It seems the timeout is caused by that coordinator is waiting for the offset partition, and our CI could be too busy to complete the assignments.
> 
> Hi @chia7712, thanks for the suggestion. I have set `offsets.topic.num.partitions` as `1` on `ClusterTestDefaults`. Hope it works fine.

Hi @chia7712, setting `offsets.topic.num.partitions` as `1` works! Do you think that we should revert `unstable.api.versions.enable` change and try again? Thanks.

**Comment by chia7712** [comment]
> Do you think that we should revert unstable.api.versions.enable change and try again? Thanks.

Yep

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Could you remove the usage of `TestUtils`? the method `subscribeAndWaitForRecords` can be rewrite easily as it seems to poll records only.

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 overall LGTM. one small comment left. please take a look.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Hi @chia7712, I removed TestUtils in `DeleteOffsetsConsumerGroupCommandIntegrationTest`. Thank you.

Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: It looks like the only place using `subscribeAndWaitForRecords` is `PlaintextAdminIntegrationTest` after this PR is merge. I will create another minor PR to refactor it. Thanks.

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: > I will create another minor PR to refactor it. Thanks.

nice!!!

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Could we rewrite it by `assertNotEquals(0, records.count())`?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: please use  try-with-resources

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: please use  try-with-resources

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for keeping updating code. a couple of comments left. Also, please use `try-with-resources` to simplify code.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Hi @chia7712, thanks for the review. I have addressed all following comments.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Could we reduce the replica number instead of increasing number of brokers? I try to avoid our CI get burned :)

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Could we use try-with-resource to close `Admin`? We don't guarantee that `createAdminClient` will auto-release the admin currently. I have filed https://issues.apache.org/jira/browse/KAFKA-16589 to discuss it. However, we should always close it.

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updated PR. two minor comments left. PTAL

**Comment by FrankYang0529** [comment]
> @FrankYang0529 thanks for updated PR. two minor comments left. PTAL

Hi @chia7712, I addressed last comments. Thanks for the review.

**Review by chia7712 — APPROVED** [review]
LGTM

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: please remove `;`

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Can we merge `withStableConsumerGroup` and `withEmptyConsumerGroup`?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Could you add tests for `GroupProtocol.CONSUMER`? see comment: https://github.com/apache/kafka/pull/15766#discussion_r1576009920

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 sorry that I check the PR again, and more comments are left. PTAL

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Can we do it in another PR?

**Comment by FrankYang0529** [comment]
> @FrankYang0529 sorry that I check the PR again, and more comments are left. PTAL

Hi @chia7712, I addressed last comments. I think we can add new test case in next PR. We can more focus on migrate to ClusterTestExtensions in this PR.

**Comment by chia7712** [comment]
> I think we can add new test case in next PR. We can more focus on migrate to ClusterTestExtensions in this PR.

Please take a look at https://github.com/apache/kafka/pull/15766#discussion_r1576009920, and we need a bit refactor but it should not a hard work. So please try to address it in this PR.

**Comment by FrankYang0529** [comment]
> > I think we can add new test case in next PR. We can more focus on migrate to ClusterTestExtensions in this PR.
> 
> Please take a look at [#15766 (comment)](https://github.com/apache/kafka/pull/15766#discussion_r1576009920), and we need a bit refactor but it should not a hard work. So please try to address it in this PR.

Update it. Thanks.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: It seems `config` is always empty, so please remove it.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: we should test both `CONSUMER` and `CLASSIC` for kraft. Also, could you initialize the consumer configs in construction. for example;
```java
    DeleteOffsetsConsumerGroupCommandIntegrationTest(ClusterInstance clusterInstance) {
        this.clusterInstance = clusterInstance;
        this.consumerConfigs = clusterInstance.isKRaftTest()
            ? Arrays.asList(Collections.singletonMap(ConsumerConfig.GROUP_PROTOCOL_CONFIG, GroupProtocol.CLASSIC.name()),
                Collections.singletonMap(ConsumerConfig.GROUP_PROTOCOL_CONFIG, GroupProtocol.CONSUMER.name()))
            : Collections.emptyList();
    }
```

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updated PR. please take a look at two comments.

**Comment by FrankYang0529** [comment]
> @FrankYang0529 thanks for updated PR. please take a look at two comments.

Hi @chia7712, thanks for the review. Updated it.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: We can change the type from `Properties` to `Map<String, Object>`. With that change we don't need to create a lot of `Properties` in each test case

Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Updated it and also added empty map for zk, or zk will not be tested.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: We can use immutable map and make `createConsumer` create a inner mutable map to collect all configs.

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 one last comment. PTAL

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java
Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Updated it. Thanks.

**Pull request merged**
